### PR TITLE
Dqn refactor

### PR DESCRIPTION
**Pull request by zhog96**

### CODE REVIEW

File: ml/dqn_encoders.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            values=all_neighbors_embeddings,
+            lengths=neighbor_node_ids.lengths
+        )
+
+        # Shape: [batch_size, max_neighbors_num, embedding_dim]
+        padded_neighbors_node_embeddings = neighbor_node_embeddings.padded_values
+
+        # 1) Shift neighbors and destinations
+        # TODO[Vladimir Baikalov]: Make shifting optional
+        # TODO[Vladimir Baikalov]: Check that it doesn't lead to gradient issues
```
Comment (NonameUntitled) [inline ml/dqn_encoders.py:61]: It is also can be removed. Afair, this behavior was checked for me during PPO testing. If it is not correct and the same TODO exists for PPO, you can just drop this issue.

File: ml/dqn_encoders.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        neighbor_node_embeddings = TensorWithMask(
+            values=all_neighbors_embeddings,
+            lengths=neighbor_node_ids.lengths
+        )
+
+        # Shape: [batch_size, max_neighbors_num, embedding_dim]
+        padded_neighbors_node_embeddings = neighbor_node_embeddings.padded_values
+
+        # 1) Shift neighbors and destinations
+        # TODO[Vladimir Baikalov]: Make shifting optional
```
Comment (NonameUntitled) [inline ml/dqn_encoders.py:60]: It is already done 😄. You can check PPO implementation and do the same here.

File: ml/dqn_encoders.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    def forward(
+            self,
+            current_node_idx: Tensor,
+            neighbor_node_ids: TensorWithMask,
+            destination_node_idx: Tensor
+    ) -> Tuple[Tensor, Tensor]:
+        raise NotImplementedError
+
+
+class TowerQNetwork(BaseQNetwork, config_name='tower_q_network'):
```
Comment (NonameUntitled) [inline ml/dqn_encoders.py:23]: Generic question: why do you need different tower encoders for PPO and DQN? I think it can be implemented as a common encoder which can be suitable for DQN, PPO, and other different approaches.

File: agents/ppo.py [hunk 4]
```diff
@@ -97,11 +97,11 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
                 bag_ids=bag_ids,
                 node_idxs=current_node_idx,
                 extra_infos=zip(
-                    current_state_value_function,
-                    neighbors_logits,
+                    current_state_value_function.detach(),
```
Comment (NonameUntitled) [inline agents/ppo.py:100]: Wait, wait, wait. What do you need  `detach` here? I thought you want to store all the information about gradient flows as well.

File: agents/dqn.py [hunk 5]
```diff
@@ -1,24 +1,132 @@
...
+                    neighbor_node_ids=neighbor_node_ids_,
+                    destination_node_idx=destination_node_idx_
+                )
+                target += self._discount_factor * neighbor_q_
+            neighbor_q = self._q_network(
+                current_node_idx=current_node_idx,
+                neighbor_node_ids=neighbor_node_ids,
+                destination_node_idx=destination_node_idx
+            )
+            loss = (target.detach() - neighbor_q[next_neighbor]) ** 2
```
Comment (NonameUntitled) [inline agents/dqn.py:132]: This is incorrect! By using `detach` you remove all the information about gradient flow. Please, change it, otherwise `loss.backward()` will not work properly. Description of a `detach` method: "Returns a new Tensor, detached from the current graph. The result will never require gradient."

File: agents/dqn.py [hunk 6]
```diff
@@ -1,24 +1,132 @@
...
+            current_node_idx_prefix=config['current_node_idx_prefix'],
+            destination_node_idx_prefix=config['destination_node_idx_prefix'],
+            neighbors_node_ids_prefix=config['neighbors_node_ids_prefix'],
+            q_network=BaseEncoder.create_from_config(config['q_network']),
+            discount_factor=config.get('discount_factor', 0.99),
+            research_prob=config.get('research_prob', 0.1),
+            bag_trajectory_memory=BaseBagTrajectoryMemory.create_from_config(config['path_memory'])
+            if 'path_memory' in config else None,
+            node_idx=config.get('node_idx', None),
+            sample_size=config.get('sample_size', 100)
```
Comment (NonameUntitled) [inline agents/dqn.py:69]: This is not consistent with the `__init__` function signature. If you are using `config.get(...)` in this function, then in the `__init__` the same variable will also have a default value, otherwise, you do it implicitly which is bad. If you don't want to have a default param in `__init__`, then please remove `.get(...)` from here.

File: agents/dqn.py [hunk 7]
```diff
@@ -1,24 +1,132 @@
...
+            current_node_idx=current_node_idx,
+            neighbor_node_ids=neighbor_node_ids,
+            destination_node_idx=destination_node_idx
         )
 
-    def forward(self, inputs):
-        for part in self._parts:
-            inputs = part(inputs)
-        return inputs
+        next_neighbor_ids = neighbor_node_ids.flatten_values[torch.argmax(next_neighbor_q, dim=1)]
```
Comment (NonameUntitled) [inline agents/dqn.py:86]: What if `torch.argmax` will return the index of the padded neighbor? Is it possible? Can you please reply to it? Also, can you please write info about shapes in order to simplify code review and further debugging?

Comment (NonameUntitled) [inline agents/dqn.py:86]: This looks strange to me. `next_neighbor_q` has shape `[batch_size, num_neighbors]`, so `torch.argmax(next_neighbor_q, dim=1)` will have shape `[batch_size]` where all the values are from the following range: `[0...max_num_neighbors - 1]`. When you call `flatten_values` you receive a tensor with shape `[all_batch_neighbors]`, so I assume there is an error in your pipeline. Can you please elaborate in the comments here?

File: agents/dqn.py [hunk 8]
```diff
@@ -1,24 +1,132 @@
...
 
 
-# TODO[Vladimir Baikalov]: Implement
-class DQNAgent(TorchAgent, config_name='dqn'):
+def _with_random_research(
+        argmax_next_neighbor: Tensor,
+        neighbor: TensorWithMask,
+        prob: float
+):
+    mask = torch.bernoulli(torch.full(argmax_next_neighbor.shape, prob)).int()
```
Comment (NonameUntitled) [inline agents/dqn.py:19]: Btw, why do you want to use Bernoulli here? Afaik, it is possible to get the exact prop by applying `torch.softmax` to scores from your `q_network`.

**Review by NonameUntitled — CHANGES_REQUESTED** [review]

File: ml/dqn_encoders.py (hunk 3 above)
Comment (zhog96) [reply ml/dqn_encoders.py:23]: There may be little difference in implementation, so may be leave two different classes?, may be refactor something during integration with env

File: agents/dqn.py (hunk 6 above)
Comment (zhog96) [reply agents/dqn.py:69]: Hm, I don't completely understand what you mean here =)

File: agents/dqn.py (hunk 7 above)
Comment (zhog96) [reply agents/dqn.py:86]: yes, it should be padded_values)

File: agents/dqn.py (hunk 8 above)
Comment (zhog96) [reply agents/dqn.py:19]: Prob is prob of random search, it doesn't concern q_values

File: ml/dqn_encoders.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+import torch
+from torch import Tensor
+
+from ml import BaseEmbedding
+from ml.encoders import TorchEncoder, TowerEncoder
+from ml.utils import TensorWithMask
+
+
+class BaseQNetwork(TorchEncoder, config_name='base_q_network'):
```
[the remark this answers is not in the document]
Comment (zhog96) [reply ml/dqn_encoders.py:12]: I thought that there may be some another inheritance in future). Ok, config_name removed.

File: agents/dqn.py (hunk 5 above)
Comment (zhog96) [reply agents/dqn.py:132]: I know what detach does) I thought that old value should be detached in dqn, may be discus later in telegram?

File: ml/dqn_encoders.py (hunk 3 above)
Comment (NonameUntitled) [reply ml/dqn_encoders.py:23]: Ok, this was just a suggestion :)

File: agents/dqn.py (hunk 8 above)
Comment (NonameUntitled) [reply agents/dqn.py:19]: Ok, but afaik, we decided to implement solution with `torch.distributions.Categorical`. Correct me if I am wrong

File: agents/dqn.py [hunk 10]
```diff
@@ -75,17 +75,20 @@ def forward(self, inputs: Dict[str, Any]) -> Dict[str, Any]:
...
             current_node_idx=current_node_idx,
             neighbor_node_ids=neighbor_node_ids,
             destination_node_idx=destination_node_idx
         )
 
-        next_neighbor_ids = neighbor_node_ids.flatten_values[torch.argmax(next_neighbor_q, dim=1)]
-        next_neighbors_idx = _with_random_research(
-            next_neighbor_ids,
+        # Shape: [batch_size]
+        best_next_neighbor_idx = neighbor_node_ids.padded_values[torch.argmax(next_neighbor_q, dim=-1)]
```
Comment (NonameUntitled) [inline agents/dqn.py:89]: What if `torch.argmax(next_neighbor_q, dim=-1)` will give you padded idx? Can you please explicitly make something like this `neighbor_node_ids.padded_values[neighbor_node_ids.mask] = -torch.inf` and only then use argmax

File: ml/dqn_encoders.py [hunk 11]
```diff
@@ -56,20 +63,36 @@ def forward(
...
+        all_state_embeddings = torch.cat(
+            [current_state_embedding_expanded, next_state_embeddings],
+            dim=-1
+        )
 
         # 2) Compute q func
-        # Shape: [batch_size, embedding_dim]
-        neighbors_q = self._ff_net.forward(shifted_destination_node_embedding)
-
-        # TODO[Vladimir Baikalov]: Probably it's a good idea to divide logits to make the distribution more smooth
```
Comment (NonameUntitled) [inline ml/dqn_encoders.py]: This line should not be removed

**Review by NonameUntitled — CHANGES_REQUESTED** [review]

File: agents/ppo.py (hunk 4 above)
Comment (zhog96) [reply agents/ppo.py:100]: If this works bad, will be fixed further

File: agents/dqn.py (hunk 7 above)
Comment (zhog96) [reply agents/dqn.py]: no, q_network returns tensor with -inf in it

File: agents/dqn.py (hunk 8 above)
Comment (zhog96) [reply agents/dqn.py:19]: done)

**Review by NonameUntitled — APPROVED** [review]
